server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java

```diff
@@ -253,15 +258,17 @@ private void putInternal(HotRodHeader header, AdvancedCache<byte[], byte[]> cach
                .putAsync(key, value, metadata)
                .thenApply(CompletableFutures.toNullFunction());
       }
-      cs.whenComplete((ce, throwable) -> handlePut(header, ce, throwable))
-            .whenComplete(span);
+      cs.whenComplete((ce, throwable) -> handlePut(header, ce, throwable, span));
    }
 
-   private void handlePut(HotRodHeader header, CacheEntry<byte[], byte[]> ce, Throwable throwable) {
+   private void handlePut(HotRodHeader header, CacheEntry<byte[], byte[]> ce, Throwable throwable,
+                          InfinispanSpan<CacheEntry<?, ?>> span) {
       if (throwable != null) {
          writeException(header, throwable);
+         span.accept(null, throwable);
       } else {
          writeSuccess(header, ce);
+         span.accept(ce, null);
```

Comment:
```suggestion
         span.complete();
```

Reply:
CompletionStage uses proper wildcard generics `? super T` so if it is just Object it would work just the same.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html#whenComplete-java.util.function.BiConsumer-

Reply:
You are more referring to implementing CompletionStage, which I agree we would want that.

Reply:
:facepalm: It is the first time I've noticed it uses `? super T`